SHOW TABLES FROM <qualified_name>

xevix opened the pull request:
Add support for `SHOW TABLES FROM db` and `SHOW TABLES FROM db.schema`. Avoids need to `USE db` or `USE db.schema` just to see the tables present.

```
D ATTACH '~/taxi.duckdb';
D SHOW TABLES FROM taxi;
┌────────────────┐
│      name      │
│    varchar     │
├────────────────┤
│ taxi_data_2019 │
│ zone_lookups   │
└────────────────┘
D SHOW TABLES FROM taxi.main;
┌────────────────┐
│      name      │
│    varchar     │
├────────────────┤
│ taxi_data_2019 │
│ zone_lookups   │
└────────────────┘
```

The existing parse/transform/bind implementation for `SHOW TABLES/DATABASES/VARIABLES` is using magic string comparisons which is a bit hard to follow, but I'm leaving that as-is to keep the scope of this PR small. If this is happening for some edge case reason in the parser or so, please let me know.

Implements https://github.com/duckdb/duckdb/discussions/16083

- [x] Tests

xevix committed:
SHOW TABLES FROM <qualified_name>.

Mytherin commented on src/function/pragma/pragma_queries.cpp:
```diff
@@ -19,19 +19,32 @@ string PragmaTableInfo(ClientContext &context, const FunctionParameters &paramet
+	string where_clause = "";
+	vector<string> where_conditions;
+	if (!database.empty()) {
+		where_conditions.push_back("database_name = '" + database + "'");
```
Can we use `StringUtil::Format` with `SQLString(database)` as parameters so that these are properly escaped?

Mytherin replied:
Can we also add a test with a database name that has quotes in it?

Mytherin commented on third_party/libpg_query/grammar/statements/variable_show.y:
```diff
@@ -21,10 +21,15 @@ VariableShowStmt:
 			{
 				PGVariableShowStmt *n = makeNode(PGVariableShowStmt);
-				n->relation = $2;
+				if ($3) {
```
It looks to me like "$2" is ignored if "$3" is set, what if we do `SHOW SOMETHING_ELSE FROM DB`? Can we at least verify that this is `SHOW TABLES FROM ...`?

Mytherin reviewed (commented):
Thanks for the PR! Looks good to me, some comments. Can we also add some tests?

xevix committed:
Clearer grammar for SHOW TABLES FROM, add tests.

xevix commented:
Thanks for the review! I've addressed the comments. Sorry for the incomplete PR, I had marked it as draft but I'll make sure to be ready for full review before opening a PR next time. Mostly wanted to confirm this was a desirable feature first.

A question on syntax, currently `USE` only allows `database` or `database.schema` as a parameter [according to the docs](https://duckdb.org/docs/stable/sql/statements/use.html), but if you actually run `USE main` to go to the main schema it will work due to the `TransformUse` [implementation](https://github.com/duckdb/duckdb/blob/223ff0a7dba7900039d5910a009247ef097fff3c/src/parser/transform/statement/transform_use.cpp#L20). I'm not sure if this is intentional, but perhaps this should be consistent with `SHOW TABLES FROM`. Either `USE` is changed to work as it states in the documentation to expect only db or db+schema, or `SHOW TABLES FROM` can accept schemas, but in either case it makes sense to update the documentation to reflect this.

I've updated the grammar, but a test is failing with verification_enabled in `test_show_tables_from.test`. If I run `SHOW TABLES FROM main` without it enabled, it properly shows 1 column, but with verification on, it seems to show a schema similar to `SHOW ALL TABLES` and I'm not sure why. I'll keep digging into this.

```
D show tables from memory.main;
┌─────────┐
│  name   │
│ varchar │
├─────────┤
│ 0 rows  │
└─────────┘
D PRAGMA enable_verification
  ;
┌─────────┐
│ Success │
│ boolean │
├─────────┤
│ 0 rows  │
└─────────┘
D show tables from memory.main;
Invalid Error:
Parsed statement differs from original result!
Original Result:
name	
VARCHAR	
[ Rows: 0]

Parsed:
database	schema	name	column_names	column_types	temporary	
VARCHAR	VARCHAR	VARCHAR	VARCHAR[]	VARCHAR[]	BOOLEAN	
[ Rows: 0]
```

Mytherin commented:
Enable verification triggers the `ToString()`, Serialize/Deserialize and Copy code paths - it is likely one of these that is not correctly supported.

As for the schema vs db issue - I agree it would be nice if `SHOW TABLES FROM` would match `USE` semantics.

xevix committed:
Amend SHOW TABLES FROM serdes and ToString().

xevix commented:
Thanks for the verification help, I've fixed it and that works.

For the `USE` issue, I think it might be a bug. You can run the statement `USE schemaname` (although this syntax is missing in the docs) and it will run, but the search path is not being changed. The same happens if you run `USE db.schemaname` (which is in the docs). If you can confirm, I'll file a bug for this. Once I know the expected behavior for `USE`, I'll change `SHOW TABLES FROM` to match.

```
DuckDB v1.3.1 (Ossivalis) 2063dda3e6
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.

D SELECT catalog_name, schema_name from information_schema.schemata WHERE catalog_name = 'memory';
┌──────────────┬─────────────┐
│ catalog_name │ schema_name │
│   varchar    │   varchar   │
├──────────────┼─────────────┤
│ memory       │ main        │
└──────────────┴─────────────┘
D CREATE TABLE foo (id INTEGER);
D SHOW TABLES;
┌─────────┐
│  name   │
│ varchar │
├─────────┤
│ foo     │
└─────────┘
D CREATE SCHEMA schema2;
D USE schema2;
D SHOW TABLES;
┌─────────┐
│  name   │
│ varchar │
├─────────┤
│ foo     │
└─────────┘
D USE main;
D SHOW ALL TABLES;
┌──────────┬─────────┬─────────┬──────────────┬──────────────┬───────────┐
│ database │ schema  │  name   │ column_names │ column_types │ temporary │
│ varchar  │ varchar │ varchar │  varchar[]   │  varchar[]   │  boolean  │
├──────────┼─────────┼─────────┼──────────────┼──────────────┼───────────┤
│ memory   │ main    │ foo     │ [id]         │ [INTEGER]    │ false     │
└──────────┴─────────┴─────────┴──────────────┴──────────────┴───────────┘
D USE memory.schema2;
D SHOW TABLES;
┌─────────┐
│  name   │
│ varchar │
├─────────┤
│ foo     │
└─────────┘
```

Mytherin commented:
`USE [schemaname]` is supposed to work (and predates support for multiple databases) - so that might be a bug in `SHOW TABLES`.

xevix committed:
Fix SHOW TABLES FROM tests.

xevix committed:
ShowRef make generate-files.

xevix committed:
Merge branch 'main' into show-tables-from-database

xevix committed:
Fix SHOW TABLES FROM tests with latest_storage.json.

Mytherin commented on src/planner/binder/tableref/bind_showref.cpp:
```diff
@@ -151,6 +155,44 @@ unique_ptr<BoundTableRef> Binder::BindShowTable(ShowRef &ref) {
+		auto schema_name = ref.schema_name;
+
+		// First, try to resolve as a schema name
+		auto schema_entry = Catalog::GetSchema(context, catalog_name, schema_name, OnEntryNotFound::RETURN_NULL);
```
Can we use `Binder::BindSchemaOrCatalog` here to figure out if something is a schema or a catalog? That also deals with detecting ambiguous references.

xevix replied:
Sure, can do. I was trying to see what `USE` did in this case, which uses a slightly different logic. Is it worth consolidating this with the Binder's approach? Or maybe out of scope for this case.


https://github.com/duckdb/duckdb/blob/59382ca326bd84009ee982657c17cef91054dd09/src/catalog/catalog_search_path.cpp#L162-L182

https://github.com/duckdb/duckdb/blob/59382ca326bd84009ee982657c17cef91054dd09/src/parser/transform/statement/transform_use.cpp#L7-L21

xevix committed:
Merge branch 'main' into show-tables-from-database

xevix committed:
Use Binder::BindSchemaOrCatalog to disambiguate db and schema.
